lib/Sema/ConstraintSystem.cpp

```diff
@@ -2883,14 +2913,45 @@ static bool diagnoseContextualFunctionCallGenericAmbiguity(
     if (genericParamInferredTypes.size() != 2)
       return false;
 
-    auto &DE = cs.getASTContext().Diags;
+    auto describeType = [&](Type argType, const PrintOptions &PO) -> std::string {
+      std::string Result;
+      llvm::raw_string_ostream OS(Result);
+
+      OS << "'";
+      argType.print(OS, PO);
+      OS << "'";
+
+      return OS.str();
+    };
+
+    llvm::SmallDenseMap<TypeBase *, std::string> descriptions;
+    llvm::StringMap<unsigned> descriptionCounts;
+    for (const auto &type : genericParamInferredTypes) {
+      auto description = describeType(type, PrintOptions());
+      descriptions[type.getPointer()] = description;
+      descriptionCounts[description] += 1;
+    }
+
     llvm::SmallString<64> arguments;
     llvm::raw_svector_ostream OS(arguments);
+
     interleave(
         genericParamInferredTypes,
-        [&](Type argType) { OS << "'" << argType << "'"; },
+        [&](Type argType) {
+          auto description = descriptions[argType.getPointer()];
+          if (descriptionCounts[description] == 1) {
+            OS << description;
+            return;
+          }
+          // Two or more types have the exact same description. Fully-qualify
+          // the types to help the user disambiguate them.
+          auto PO = PrintOptions();
+          PO.FullyQualifiedTypes = true;
```

Comment:
I think we need to solve this problem in a diagnostic agnostic way, this logic should instead go into DiagnosticEngine where we transform types passed as arguments and if nominal types descriptions match we should print them in fully qualified form.